mesonbuild/interpreter/type_checking.py

```diff
@@ -29,6 +30,87 @@
     from ..interpreterbase import TYPE_var
     from ..interpreterbase.decorators import FeatureCheckBase
 
+
+def _types_feature_change_validator(
+        since: bool,
+        types: T.Optional[T.Dict[T.Type, T.Union[str, T.Tuple[str, str]]]]
+        ) -> T.Callable[[object, str], T.Iterable[FeatureCheckBase]]:
+    """Check for types and emit FeatureCheckBase checks
+
+    :Note: this behaves differently than the value checker in regards to
+    dictionaries, as it looks at the values, while the value check looks at the
+    keys.
+
+    :param types: A mapping of types to versions or (version, extra_message)
+    :return: A function suitable to pass to :attribute:`KwargInfo.feature_validator`
+    """
+
+    change = FeatureNew if since else FeatureDeprecated
+
+    def validator(to_validate: object, prefix: str) -> T.Iterable[FeatureCheckBase]:
+        for type_, version in types.items():
+            message: T.Optional[str]
+            if isinstance(version, tuple):
+                version, message = version
+            else:
+                message = None
+
+            # Handle scaler types (str, int, etc) *and* dictionaries and lists
+            # We do sometimes deprecate passing lists and dicts.
+            if type_ in {dict, list} or not isinstance(to_validate, (dict, list)):
+                if isinstance(to_validate, type_):
+                    yield change(f"{prefix} of type {type_.__name__}", version, message)
+            elif isinstance(to_validate, dict):
+                for v in to_validate.values():
+                    if isinstance(v, type_):
+                        yield change(f"{prefix} of type {type_.__name__}", version, message)
```

Comment:
Can't this yield the same `Feature*` if the type in question is used multiple times in the same dict?

Maybe this should be `if any(...)` instead of `for v in ...: if isinstance(...)`.